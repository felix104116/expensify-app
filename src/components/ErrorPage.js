import React from 'react';
import { Link } from 'react-router-dom';
const ErrorPage = () => (
    <div>
        <p>Error 404</p>
        <Link to="/">Go Home</Link>
    </div>
);
export default ErrorPage;