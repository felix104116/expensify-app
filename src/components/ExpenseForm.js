import React from 'react';
import {connect} from 'react-redux';
import {addExpense} from '../actions/expenses';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

//const date = new Date();
const date = moment();

export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount/100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calenderFocused:false,
            error: '',
        };
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(() => ({amount}))
        }
    };
    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(() => ({createdAt}))
        }
    };
    onFocusChange = ({focused}) => {
        this.setState(() => ({calenderFocused:focused}))
    };
    handleAddExpense = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(() => ({error:'Please provide a description and an amount.'}));
        }
        else{
            this.setState(() => ({error:''}));
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: parseFloat(this.state.amount,10)*100,
                createdAt: this.state.createdAt.valueOf(),
            });
        }
    }
    render(){
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.handleAddExpense}>
                    <input type="text" placeholder="Description" autoFocus="true" value={this.state.description} onChange={this.onDescriptionChange}/>
                    <input type="text" placeholder="Amount" value={this.state.amount} onChange={this.onAmountChange}/>
                    <textarea value={this.state.note} placeholder="Add a note for your expense (optional)" onChange={this.onNoteChange}/>
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calenderFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}