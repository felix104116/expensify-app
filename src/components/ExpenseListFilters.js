import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null,
    };
    onDatesChange = ({startDate,endDate}) =>{
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocused) => {
        this.setState({calendarFocused})
    }
    onSortByText = (e) => this.props.setTextFilter(e.target.value);
    onSortChange = (e) => {
        if(e.target.value === 'date')
            this.props.sortByDate();
        else
            this.props.sortByAmount();
    }
    render(){
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onSortByText}/>
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                {this.props.filters.sortBy === 'date' &&
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    startDateId="startDate"
                    endDateId="endDate"
                    endDate={this.props.filters.endDate}
                    focusedInput={this.state.calendarFocused}
                    onDatesChange={this.onDatesChange}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                }
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return{
        filters:state.filters
    };
}

const mapDispatchToProps = (dispatch,props) => {
    return {
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate))
    };
}

export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters);