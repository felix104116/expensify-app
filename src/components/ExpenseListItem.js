import React from 'react';
import { connect } from 'react-redux';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

export const ExpenseListItem = ({id,description,amount,createdAt}) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>costs {numeral(amount/100).format('$0,0.00')} and was created on {moment(createdAt).format('MMM Do,YYYY')}</p>
    </div>  
);
export default connect()(ExpenseListItem);