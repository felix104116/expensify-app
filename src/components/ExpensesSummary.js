import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import {history} from '../routers/AppRouter';

//Expense count n'est pas mis Ã  jour
export const ExpensesSummary = ({expenseCount,expenseTotal}) => {
    const expenseWord = expenseCount > 1 ? 'expenses' : 'expense';
    const formatedExpenseTotal = numeral(expenseTotal/100).format('$0,0.00');
    return(
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Viewing <span>{expenseCount}</span> {expenseWord} totaling <span>{formatedExpenseTotal}</span></h1>
                <button className="bigButton" onClick={() => history.push('/create')}>Add Expense</button>
            </div>
        </div>
    )
}

const mapStateToProps = (state) => {
    const expenses = getExpenses(state.expenses,state.filters);
    return {
        expenseCount:expenses.length,
        expenseTotal: getExpensesTotal(expenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary);