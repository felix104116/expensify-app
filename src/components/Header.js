import React from 'react';
import {  Link } from 'react-router-dom';
import {startLogout} from '../actions/auth';
import { connect } from 'react-redux';
export const Header = ({startLogout}) => (
    <header className="header__navbar">
        <div className="content-container">
            <div className="header__content">
                <Link className="header__title" to="/dashboard">
                    <h1 className="page-header__title">Expensify</h1>
                </Link>
                <button className="header__logout_button" onClick={startLogout}>Logout</button>
            </div>
        </div>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined,mapDispatchToProps)(Header);