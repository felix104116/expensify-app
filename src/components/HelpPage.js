import React from 'react';

const HelpPage = () => (
    <div>
        <p>This is the help page component</p>
    </div>
);
export default HelpPage;