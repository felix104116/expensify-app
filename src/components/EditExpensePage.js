import React from 'react';
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id,expense);
        this.props.history.push('/');
    };
    onRemove = () => {
        this.props.startRemoveExpense({id:this.props.expense.id});
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <div className="content-container">
                    <div className="page-header">
                        <h1 className="page-header__title">Edit Page</h1>
                    </div>
                </div>
                <div className="content-container">
                <ExpenseForm 
                onSubmit={this.onSubmit}
                expense={this.props.expense}/>
                <div className="remove-container">
                    <button className="bigButton button--secondary" onClick={this.onRemove}>Remove Expense</button>
                </div>
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch,props) => ({
        startEditExpense: (id,expense) => dispatch(startEditExpense(id,expense)),
        startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

const mapStateToProps = (state, props) => {
    return{
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
}
export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);