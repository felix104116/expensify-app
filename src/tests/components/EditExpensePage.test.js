import {EditExpensePage} from '../../components/EditExpensePage';
import React from 'react';
import {shallow} from 'enzyme';
import expenses from '../fixtures/expenses';

let editExpense, startRemoveExpense, history, wrapper;


beforeAll(() => {
    editExpense = jest.fn();
    startRemoveExpense = jest.fn();
    history = {
        push: jest.fn()
    };
    wrapper = shallow(
        <EditExpensePage 
        editExpense={editExpense} 
        startRemoveExpense={startRemoveExpense} 
        history={history} 
        expense={expenses[0]}
        />
        );
})

test('should render EditExpensePage',() => expect(wrapper).toMatchSnapshot());

test('should remove a component', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith({id:expenses[0].id});
})

test('should handle edit expense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenCalledWith(expenses[0].id,expenses[0]);
})