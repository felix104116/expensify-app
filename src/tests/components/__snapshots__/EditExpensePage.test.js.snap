// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <div
    className="content-container"
  >
    <div
      className="page-header"
    >
      <h1
        className="page-header__title"
      >
        Edit Page
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 95000,
          "createdAt": 345600000,
          "description": "Rent",
          "id": "1",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <div
      className="remove-container"
    >
      <button
        className="bigButton button--secondary"
        onClick={[Function]}
      >
        Remove Expense
      </button>
    </div>
  </div>
</div>
`;
